include/seahorn/Expr/ExprRewriteRule.hh

```diff
@@ -151,7 +152,45 @@ struct BoolOpRewriteRule : public ExprRewriteRule {
     if (isOpX<NEG>(exp)) {
       Expr neg = exp->arg(0);
       if (isOpX<NEG>(neg)) {
-        return {neg->arg(0), rewrite_status::RW_2};
+        return {neg->arg(0), rewrite_status::RW_1};
+      }
+    }
+    return {exp, rewrite_status::RW_SKIP};
+  }
+};
+
+// for select and store
+struct ArrayRewriteRule : public ExprRewriteRule {
+  ArrayRewriteRule(ExprFactory &efac) : ExprRewriteRule(efac) {}
+  ArrayRewriteRule(const ArrayRewriteRule &o) : ExprRewriteRule(o) {}
+
+  rewrite_result operator()(Expr exp) {
+    if (!isOpX<STORE>(exp) && !isOpX<SELECT>(exp)) {
+      return {exp, rewrite_status::RW_SKIP};
+    }
+    if (isOpX<SELECT>(exp)) {
+      Expr arr = exp->arg(0);
+      Expr idx = exp->arg(1);
+      // select(ite(i, a, b), v) => ite(i, select(a, v), select(b, v))
+      if (isOpX<ITE>(arr)) {
+        Expr i = arr->arg(0);
+        Expr t = arr->arg(1);
+        Expr e = arr->arg(2);
+        Expr new_t = op::array::select(t, idx);
+        Expr new_e = op::array::select(e, idx);
+        return {mk<ITE>(i, new_t, new_e), rewrite_status::RW_2};
```

Comment:
overall, this is potentially very expensive in our application. It will create a lot of intermediate expressions while `select` is being pushed down. It would be more effective to implement `push_select_over_ite` method that pushes select to the bottom of the `ite` tree in one operation.